### PR TITLE
improve error message when running into an unknown target dependency

### PR DESCRIPTION
**Pull request by tomerd**
motivation: help users reason about erros when encountering an unknown dependency

changes:
* include the dependency location in the error string
* refactor tests to cover more use cases

rdar://113411527

### CODE REVIEW

File: Sources/PackageLoading/ManifestLoader+Validation.swift [hunk 1]
```diff
@@ -328,3 +321,29 @@ extension TargetDescription {
...
+                case .local(let path):
+                    return path.pathString
+                case .remote(let url):
+                    return url.absoluteString
+                }
+            case .registry:
+                return .none
+            }
+        }() {
+            description += " (\(locationsString))"
```
Comment (neonichu) [inline Sources/PackageLoading/ManifestLoader+Validation.swift:344]: I think we need to workshop this a bit more

> valid dependencies are: 'foo (http://scm.com/org/foo)'

This looks extremely odd to me, how about something like:

> valid dependencies are: 'foo' (at 'http://scm.com/org/foo')

Comment (tomerd) [reply Sources/PackageLoading/ManifestLoader+Validation.swift:344]: maybe `at` for path, and `from` for SCM?

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by neonichu** [comment]
BTW, since this is to provide additional info, I wonder whether we should also reach back into mirrors and show both locations for them? Or at least provide some way to see that info? I guess the later would probably be a separate change...

**Comment by tomerd** [comment]
> BTW, since this is to provide additional info, I wonder whether we should also reach back into mirrors and show both locations for them? Or at least provide some way to see that info? I guess the later would probably be a separate change...

the mirrors information is not currently tracked in a way that is easy to access, so seems like a bigger change

**Review by neonichu — APPROVED** [review]

**Pull request merged**